---
title: "DE analysis of Depmap melanoma bulk RNAseq counts"
author: "Theo Killian"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r, echo = FALSE, results = 'asis'}
# Set global knitr options
knitr::opts_chunk$set(
  collapse = TRUE,
  comment  = "#>",
  warning  = FALSE,
  message  = FALSE
)

# Load required packages
suppressPackageStartupMessages({
  library(dplyr)
  library(tidyr)
  library(tibble)
  library(ggplot2)
  library(ggrepel)
  library(DESeq2)
  library(stringr)
  library(RColorBrewer)
  library(pheatmap)
  library(readxl)
  library(readr)
  library(writexl)
  library(biomaRt)
  library(tximport)
  library(txdbmaker)
  library(GenomicFeatures)
  library(SummarizedExperiment)
  library(janitor)
  # library(gridExtra)
  library(DT)
})

# set project name
project_name <- "depmap_melanoma_deseq2"

# Set seed for reproducibility
set.seed(123)
```

# Introduction

Bulk RNA-seq was performed to investigate gene expression differences under
different conditions. In this report, we describe how we processed the data,
performed differential gene expression (DE) analysis, and generated diagnostics
and visualizations using the [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)
Bioconductor package.

### Analysis Workflow

1) Load Count Data and Metadata
1) Initial Filtering
1) Use `DESeqDataSet` to build a DESeq2 object
1) QC and Exploratory Analysis (PCA, dispersion plots, etc.)
1) Perform Differential Expression
1) Annotation and Result Summaries
1) Visualizations (MA, Volcano and Count Plots and Expression Heatmap)
1) Save Outputs

**NOTE: my bioinformatic interpretations will be displayed in bold**

## Data Preparation

In this section, we load raw count data and sample metadata, and we apply
minimal filtering (e.g., removing genes with zero counts across all samples).

```{r warning=FALSE, message=FALSE}
## eval=FALSE, 
readr::read_csv("./data/OmicsDefaultModelProfiles.csv") %>%
  dplyr::rename_with(janitor::make_clean_names) %>%
  dplyr::rename(depmap_id = model_id) %>%
  dplyr::select(-profile_type) %>%
  as.data.frame() -> map_depmap_id_to_patient_id

## latest metadata, filter for melanoma datasets only
readr::read_csv("./data/Model.csv") %>%
  dplyr::rename(depmap_id = names(.)[1]) %>%
  dplyr::select(-c(PatientSubtypeFeatures, TissueOrigin:PublicComments,
                   PublicComments, HCMIID:COSMICID, contains("Source"))) %>%
  dplyr::select(depmap_id, CCLEName, everything()) %>%
  dplyr::rename_with(janitor::make_clean_names) %>%
  dplyr::filter(grepl("Melanoma", oncotree_primary_disease)) %>%
  as.data.frame() -> meta_data
# View(meta_data)

## latest count data
readr::read_csv("./data/OmicsExpressionTranscriptsExpectedCountProfile.csv") %>%
  dplyr::rename(profile_id = names(.)[1]) %>%
  dplyr::left_join(map_depmap_id_to_patient_id, by = "profile_id") %>%
  dplyr::filter(depmap_id %in% meta_data$depmap_id) %>%
  dplyr::select(-profile_id) %>%
  tidyr::pivot_longer(cols = -depmap_id,
                      names_to = "gene_transcript",
                      values_to = "count") %>%
  dplyr::mutate(gene_name = stringr::str_extract(gene_transcript, "^[^\\s]+")) %>%
  dplyr::group_by(depmap_id, gene_name) %>%
  dplyr::summarise(count = sum(as.numeric(count)), .groups = "drop") %>%
  tidyr::pivot_wider(names_from = gene_name, values_from = count) %>%
  tibble::column_to_rownames("depmap_id") %>%
  t() %>% as.data.frame() -> cts

View(cts[1:10, 1:10])
dim(cts)

## make metadata consistent with count matrix
meta_data %>%
  dplyr::arrange(depmap_id) %>% 
  dplyr::filter(depmap_id %in% colnames(cts)) -> meta_data2
dim(meta_data2)

saveRDS(meta_data2, file = "./data/new_depmap_melanoma_metadata.rds")
saveRDS(cts, file = "./data/new_depmap_melanoma_counts.rds")
```


<!-- ### Create Gene Count Matrix using tximport -->

<!-- Our transcript counts derived from Salmon, need to be aggregated by genes. To  -->
<!-- perform this task, we import and annotate transcript-level counts from Salmon -->
<!-- and convert them into a gene-level count matrix using the R package [tximport](https://bioconductor.org/packages/release/bioc/html/tximport.html). -->
<!-- We also derive a sample metadata table directly from the count file names, which -->
<!-- we will use for differential expression analysis with `DESeq2`. -->

<!-- ```{r make_tx2gene, eval=FALSE, warning=FALSE, message=FALSE} -->
<!-- # Create a TxDb object -->
<!-- txdb <- txdbmaker::makeTxDbFromGFF("./genome/genome.filtered.gtf", format = "gtf") -->

<!-- # Get transcript-to-gene mapping -->
<!-- AnnotationDbi::select( -->
<!--   txdb, keys = keys(txdb, "TXNAME"), -->
<!--   columns = "GENEID", -->
<!--   keytype = "TXNAME") %>%  -->
<!--   as.data.frame() -> tx2gene -->

<!-- readr::write_tsv(tx2gene, file = "./data/tx2gene_mouse.tsv") -->
<!-- ``` -->

<!-- ```{r load_files, warning=FALSE, message=FALSE} -->
<!-- tx2gene <- readr::read_tsv(file = "./data/tx2gene_mouse.tsv") -->

<!-- # Path to quant.sf files -->
<!-- files <- list.files("./data/counts/", pattern = "quant.sf", full.names = TRUE) -->

<!-- # Sample names -->
<!-- sample_names <- sub(".*/|_quant.sf", "", files) -->
<!-- names(files) <- sample_names -->

<!-- # Import counts -->
<!-- txi <- tximport(files, type = "salmon", tx2gene = tx2gene, countsFromAbundance = "no") -->
<!-- cts <- as.matrix(txi$counts) -->
<!-- cts <- round(cts) -->
<!-- storage.mode(cts) <- "integer" ## coerce to integer so counts play nice with DESeq2 -->
<!-- cts <- as.data.frame(cts) -->
<!-- # cts <- as.data.frame(cts) %>%  -->
<!-- #   dplyr::select(-contains("PD1_rep1")) -->
<!-- names(cts) <- gsub("results_star_salmon_|_quant\\.sf", "", names(cts)) -->

<!-- ## make the metadata table directly out of the column names of the count data -->
<!-- data.frame(samples = names(cts), -->
<!--            condition = gsub("_.*", "", names(cts)), -->
<!--            replicate = gsub("^[^_]*_", "", names(cts)), -->
<!--            row.names = names(cts)) %>%  -->
<!--   # dplyr::filter(!grepl("PD1_rep1",  samples)) %>% # |PD1_rep2 -->
<!--   as.data.frame()-> meta_data -->

<!-- # Display a preview of metadata -->
<!-- meta_data %>% DT::datatable(options = list(pageLength = 12)) -->
<!-- ``` -->

## Filter Out Genes with Zero Counts

**NOTE: genes with all zero counts are of no use to us, so we filter them out of the analysis.**

```{r}
dim(cts) # Before filtering
keep <- rowSums(cts) > 0
cts  <- cts[keep, ]
dim(cts) # After filtering
```

## Creating a DESeqDataSet

We construct a `DESeqDataSet` object and specify the design formula. The design
formula `~ condition + replicate` indicates that we are modeling gene expression
differences by condition, while accounting for batch effects.

**NOTE: This DE comparison is "relative to the IgG condition. This means that genes with positive log fold change display higher average expression in IgG samples, whereas genes with negative log fold change are higher average expression in PD1 samples**

```{r dds, warning=FALSE, message=FALSE}
DESeqDataSetFromMatrix(
  countData = cts,
  colData   = meta_data,
  design    = ~ condition + replicate) -> dds

# Quick check of the DESeqDataSet
# dds
```

## Differential Expression Analysis

We run the main DESeq function, which handles size factor estimation, dispersion
estimation, and fitting of the negative binomial generalized linear model.

```{r keep, warning=FALSE, message=FALSE}
dds <- DESeq(dds)
# resultsNames(dds)
```

## Data exploration and quality assessment (QC)

### Principal components analysis (PCA)

**NOTE: We don't see strong separation between the IgG and PD1 samples, indicating that there is a lot of inter-condition heterogeneity, and that the biological effect between the conditions is not as strong as sample heterogeneity... at least in the first two principal components/PCs.**

```{r PCA_rld, warning=FALSE, message=FALSE}
rld_dds <- rlog(dds, blind=FALSE)
pcaData_rld <- plotPCA(rld_dds, intgroup = c("condition", "replicate"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData_rld, "percentVar"))

ggplot(pcaData_rld, aes(PC1, PC2, color = condition, shape = replicate)) +
  geom_point(size = 3) +
  geom_text_repel(aes(label = name), max.overlaps = 15) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  theme_bw()
```

<!-- We use two transformations recommended by DESeq2: -->

<!-- 1) vst(): variance stabilizing transformation -->
<!-- 1) rlog(): regularized log transformation -->

<!-- Both produce stabilized data for PCA and other downstream analyses. Below, we visualize the first two principal components. -->

<!-- ```{r PCA_vst, warning=FALSE, message=FALSE} -->
<!-- vst_dds    <- vst(dds, blind=FALSE) -->
<!-- pcaData_vst <- plotPCA(vst_dds, intgroup = c("condition", "replicate"), returnData = TRUE) -->
<!-- percentVar <- round(100 * attr(pcaData_vst, "percentVar")) -->

<!-- ggplot(pcaData_vst, aes(PC1, PC2, color = condition, shape = replicate)) + -->
<!--   geom_point(size = 3) + -->
<!--   geom_text_repel(aes(label = name), max.overlaps = 15) + -->
<!--   xlab(paste0("PC1: ", percentVar[1], "% variance")) + -->
<!--   ylab(paste0("PC2: ", percentVar[2], "% variance")) + -->
<!--   coord_fixed() + -->
<!--   theme_bw() -->
<!-- ``` -->

### Dispersion Plot

The dispersion estimates are critical for modeling variance in RNA-seq data. The
plot below shows the raw gene-wise estimates (black points), the fitted
dispersion trend (red line), and final shrunk estimates (blue open circles, if
any are outliers).

```{r}
plotDispEsts(dds)
```

### Size factors

Size factors are a method of normalizing used by the *DESeq* function to
normalize the data in terms of sequencing depth. Size factor is the median ratio
of the sample over a pseudosample: for each gene, the geometric mean of all
samples. Size factors account for differences in sequencing depth are typically
centered around 1 (indicating comparable sequencing depth).

**NOTE: Most of the size factors vary about 1.0, which is what we want to see. This means that there doesn't appear to be extreme differences in numbers of reads between**

```{r size_factor, warning=FALSE, message=FALSE}
sf_data <- data.frame(sample = rownames(dds@colData),
                      size_factor = sizeFactors(dds),
                      condition = dds$condition,
                      replicate = dds$replicate)

ggplot(sf_data, aes(x = sample, y = size_factor, fill = condition)) +
  geom_bar(stat = "identity") +
  geom_hline(yintercept = mean(sf_data$size_factor),
             color = "red", linetype = "dashed") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  ggtitle("Size Factors by Sample")
```

## Obtaining Results

### DE Results for a Specific Contrast

Below, we extract results comparing between the two conditions.

**NOTE: We are treating "IgG" as the normative condition, so LFC is calculated relative to mean expression of these samples.**

```{r}
res <- results(dds, contrast = c("condition", "IgG", "PD1"), alpha = 0.05)

# Sort by p-value
res <- res[order(res$pvalue), ]

# Summary
summary(res)
```

### Results Annotation

We annotate the DE results with mouse `Ensembl` and `Entrez` IDs by accessing
the [BioMart](https://bioconductor.org/packages/release/bioc/html/biomaRt.html)
database.

```{r biomart, eval=FALSE, warning=FALSE, message=FALSE}
## annotations available in Biomart database
# listAttributes(mart = useDataset("mmusculus_gene_ensembl", useMart("ensembl"))) %>% 
#   as.data.frame() -> atr
# View(atr)

## annotate DE results with Biomart database
## NOTE: rerun this code if you need to regenerate the gene annotation file
getBM(attributes = c("ensembl_gene_id", "entrezgene_id", "external_gene_name"),
      mart = useDataset("mmusculus_gene_ensembl", useMart("ensembl"))) %>%
  dplyr::rename(ensembl_id = ensembl_gene_id,
                entrez_id = entrezgene_id,
                gene_name = external_gene_name) %>%
  dplyr::filter(stringr::str_length(gene_name) > 1,
                !duplicated(ensembl_id)) %>%
  saveRDS(file = "./data/mouse_biomart.rds")
```

We merge the DE results with the normalized counts, and annotate the Ensembl IDs
with their gene symbols, using the table we accessed from Biomart.

```{r}
mouse_biomart <- readRDS(file = "./data/mouse_biomart.rds")
merge(as.data.frame(res), as.data.frame(counts(dds, normalized = TRUE)),
      by = "row.names", sort = FALSE) %>%
  dplyr::rename(ensembl_id = names(.)[1]) %>%
  dplyr::left_join(mouse_biomart, by = "ensembl_id") %>%
  dplyr::select(gene_name, entrez_id, ensembl_id, everything()) %>%
  as.data.frame() -> res_data
```

### DE QC Metrics

The following subsections display plots that help interpret the performance of
the DE comparison and various internal thresholds that were used.

#### Independent Filtering

*DESeq2* performs independent filtering by default using the mean of normalized
counts as a filter statistic. A threshold on the filter statistic (first value)
is found which optimizes the number of adjusted p-values lower than significance
level alpha. The adjusted p-values for the genes which do not pass the filter
threshold are set to NA. The results also returns the mean of normalized counts
(second value).

```{r filterThreshold1}
cat("Filter thresh. val. and mean of norm. counts \n")
metadata(res)$filterThreshold
```

#### Plot of sample rejections vs filter quantiles

The *filterThreshold* returns the threshold chosen (vertical line in the plots
below) by the *DESeq2* analysis of the lowest quantile of the filter for which
the number of sample rejections is within 1 residual standard deviation to the
peak of a curve fit to the number of rejections over the filter quantiles. The
following diagnostic plot shows the number of rejected samples (y-axis) plotted
against quantiles of filter (x-axis).

```{r rejections_vs_quantiles1}
par(mfrow = c(1, 1))
plot(metadata(res)$filterNumRej, type = "b", main = "",
     xlab = "Quantiles of filter", ylab = "Number of rejections")
lines(metadata(res)$lo.fit, col = "red")
abline(v = metadata(res)$filterTheta)
```

#### Histogram of frequency of p-values of results

The following plot shows the number of frequency of counts (y-axis) against
p-values between 0 and 1 (x-axis).

```{r pval_freq1}
par(mfrow = c(1, 1))
hist(res$pvalue, col = "lavender", xlab = "p-values", main = "")
```

### Numbers of Significant DE genes

*How many significant DE genes upregulated relative to the "IgG" condition and downregulated in PD1 condition?*

```{r}
res_data %>% 
  dplyr::filter(padj < 0.05,
                log2FoldChange > 0) %>% 
  nrow()
```

*How many significant DE genes downregulated relative to the "IgG" condition and upregulated in PD1 condition?*

```{r}
res_data %>% 
  dplyr::filter(padj < 0.05,
                log2FoldChange < 0) %>% 
  nrow()
```

### Interactive Table of Results

```{r warning=FALSE, message=FALSE}
res_data %>%
  dplyr::select(-c(contains("_rep"), contains("_id"), lfcSE, stat)) %>%
  dplyr::mutate(
    log2FoldChange = round(log2FoldChange, 3),
    baseMean       = round(baseMean, 1),
    pvalue         = signif(pvalue, 3),
    padj           = signif(padj, 3)) %>% 
  DT::datatable(
  options = list(pageLength = 10),
  caption = "Differential Expression Results: ")
```

### MA Plot

An MA plot (log ratio vs. mean average) visualizes the overall changes in gene
expression across the normalize count quartiles.

```{r}
plotMA(res, alpha = 0.05, main = "")
```

### Volcano Plot

A Volcano plot (log2 fold change vs -log10 p-value) helps identify genes that
display large magnitude changes and high significance.

```{r warning=FALSE, message=FALSE}
res_data %>%
  dplyr::filter(!is.na(pvalue)) %>%
  dplyr::mutate(
    negLogP = -log10(pvalue),
    sig = ifelse(padj < 0.05 & abs(log2FoldChange) > 1,
                 "Significant", "Not Significant")) -> volcano_data
volcano_data %>%
  ggplot(aes(x = log2FoldChange, y = negLogP, color = sig)) +
  geom_point(size = 0.75, alpha = 0.75) +
  geom_text_repel(data = subset(volcano_data, sig == "Significant"),
                  aes(label = gene_name), size = 3, max.overlaps = Inf,
                  box.padding = 0.3, point.padding = 0.3, segment.color = "grey50") +
  scale_color_manual(values = c("Significant" = "red", "Not Significant" = "black")) +
  geom_hline(yintercept = -log10(0.05), color = "blue", linetype = "dashed") +
  geom_vline(xintercept = c(-1, 1), color = "blue", linetype = "dashed") +
  theme_bw() +
  ggtitle("Volcano Plot")
```

### Count plots

Count plots are created for the top 5 DE genes for each comparison:

```{r}
res_data %>%
  dplyr::slice(1:5) %>%
  dplyr::select(gene_name, contains("IgG"), contains("PD1")) %>%
  tidyr::gather(key = "sample", value = "count", -gene_name) %>%
  dplyr::mutate(condition = dplyr::case_when(
                  grepl("IgG", sample) ~ "IgG",
                  grepl("PD1", sample) ~ "PD1"),
                genes = factor(gene_name, levels = unique(gene_name))) %>%
  ggplot(aes(x = sample, y = log10(count), color = condition)) +
  geom_point() +
  theme(axis.text.x = element_blank()) +
  ggtitle(paste0("Normalized gene counts of top 5 DE genes")) +
  facet_grid(~gene_name)
```

### Heatmap

#### Top 20 DE Genes

```{r}
data.frame(condition = meta_data$condition,
           row.names = names(cts)) -> sample_col

res_data %>%
  dplyr::filter(!is.na(gene_name)) %>%
  dplyr::slice(1:20) %>%
  dplyr::select(gene_name, tidyr::contains("IgG"), tidyr::contains("PD1")) %>%
  tibble::column_to_rownames(var = "gene_name") %>%
  mutate(across(everything(), ~ {
    log_vals <- log10(.)
    replace(log_vals, is.infinite(log_vals), 0)})) %>%
  pheatmap::pheatmap(
    annotation_col = sample_col,
    cluster_rows = FALSE,
    cluster_cols = FALSE,
    show_colnames = FALSE,
    main = paste0("Log10 normalized gene counts of top 20 DE genes"))
```
#### Top 200 DE Genes

```{r fig.height=10}
res_data %>%
  dplyr::filter(!is.na(gene_name),
                baseMean > 100) %>%
  dplyr::slice(1:50) %>%
  dplyr::select(gene_name, tidyr::contains("IgG"), tidyr::contains("PD1")) %>%
  tibble::column_to_rownames(var = "gene_name") %>%
  mutate(across(everything(), ~ {
    log_vals <- log10(.)
    replace(log_vals, is.infinite(log_vals), 0)})) %>%
  pheatmap::pheatmap(
    annotation_col = sample_col,
    # cluster_rows = FALSE,
    # cluster_cols = FALSE,
    show_colnames = FALSE,
    # show_rownames = FALSE,
    main = paste0("Log10 normalized gene counts of top 50 DE genes"))
```

## Saving Outputs

The results are saved as `.rds` and `.xlsx` files for downstream use:

```{r write_data, warning=FALSE, message=FALSE}
# Save DE result object
saveRDS(res_data, file = paste0("./results/", project_name, "_DE_results_",
                                Sys.Date(), ".rds"))

# Save an Excel file
write_xlsx(res_data, path = paste0("./results/", project_name, "_DE_results_",
                                   Sys.Date(), ".xlsx"))
```

## Conclusions and Next Steps

In this analysis, we:

1) Filtered out genes with zero counts.
1) Normalized raw counts to account for varying library sizes.
1) Examined PCA plots, dispersion estimates, and size factors for quality control.
1) Identified significantly differentially expressed genes, focusing on log2 fold changes and adjusted p-values.
1) Annotated results with gene IDs using the Biomart database.
1) Generated various plots illustrating the DE genes between conditions.

Next steps after this DE analysis could include:

1) Functional Analysis: GO/KEGG enrichment of significant genes.
1) Validation: qPCR or other orthogonal methods for top candidate genes.
1) Further QC: Checking for batch effects, outliers, etc.

*Session Info*

```{r sessionInfo}
sessionInfo()
```

*References:*

Love, M.I., Huber, W., Anders, S. (2014) "Moderated estimation of fold change
and dispersion for RNA-seq data with DESeq2." Genome Biology, 15:550.
10.1186/s13059-014-0550-8

Anders, Simon, and Wolfgang Huber. 2010. "Differential Expression Analysis for
Sequence Count Data." Genome Biology 11:R106.
http://genomebiology.com/2010/11/10/R106.

Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W, Smyth GK (2015). "limma powers
differential expression analyses for RNA-sequencing and microarray studies."
Nucleic Acids Research, 43(7), e47. doi: 10.1093/nar/gkv007.

Van den Berge & Perraudeau *et al* "Observation weights unlock bulk
RNA-seq tools for zero inflation and single-cell applications" *Genome Biology* (2018)
[doi: 10.1186/s13059-018-1406-4](https://doi.org/10.1186/s13059-018-1406-4)